Service behaviour not correct when ESIPP annotation is toggled or deleted

girishkalele opened the pull request:
Fixes #31649 

GCE controller: Correctly add/remove http health checks based on whether the annotation was added or removed

kube-proxy: Create/Delete XLB local-endpoints balancer chain and change jump actions from KUBE-FW- chain based on the annotation.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31828)

<!-- Reviewable:end -->

Commit:
Fix kube-proxy logic to change iptables chains when ESIPP is turned on or off

Commit:
GCE Cloud Provider - correctly handle toggling of ESIPP on or off

girishkalele commented on pkg/controller/service/servicecontroller.go:
```diff
@@ -422,10 +422,15 @@ func (s *serviceCache) delete(serviceName string) {
 
 func (s *ServiceController) needsUpdate(oldService *api.Service, newService *api.Service) bool {
-	if !wantsLoadBalancer(oldService) && !wantsLoadBalancer(newService) {
+	oldServiceNeedsLoadBalancer := wantsLoadBalancer(oldService)
```
This is going to change the behaviour of service controller to periodically call EnsureLoadBalancer even when there is no change in the service.

bprashanth replied:
Make sure this doesn't have adverse effects. Create a cluster with a service, leave it idle for a few days, make sure exactly 0 cloud provider mutation calls are made. 

Even with that you can only test gce. Can you confirm that it is OK for other CPs? (in theory it should be, but I don't know for sure)

bprashanth replied:
Just to clarify I'm not against doing just, just conveying my caution so you verify the worst case doesn't happen

girishkalele replied:
I agree - I labelled the commit as Experimental - my fix should work without the periodic reconciliation, it will just be edge triggered when a create/update/delete is detected.

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -1431,6 +1447,18 @@ func (gce *GCECloud) deleteTargetPool(name, region string, hc *compute.HttpHealt
+			return err
+		}
+	} else {
+		if err := gce.DeleteHttpHealthCheck(name); err == nil {
```
can you clarify in a comment when/if we come through and cleanup this leak in the case of an error?

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -710,6 +710,34 @@ func (gce *GCECloud) EnsureLoadBalancer(clusterName string, apiService *api.Serv
+	var hcToCreate *compute.HttpHealthCheck
+	hcExisting, _ := gce.GetHttpHealthCheck(loadBalancerName)
+	if path, healthCheckNodePort := apiservice.GetServiceHealthCheckPathPort(apiService); path != "" {
+		if hcExisting == nil {
```
this is dangerous, you could've receive any error and you're concluding that the hc doesn't exist so you need to recreat the targetPool. Suggest checking for explicitly `isHTTPErrorCode(err, http.StatusNotFound)`

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -710,6 +710,34 @@ func (gce *GCECloud) EnsureLoadBalancer(clusterName string, apiService *api.Serv
+			tpNeedsUpdate = true
+		}
+		glog.V(4).Infof("service %v needs health checks on :%d/%s)", apiService.Name, healthCheckNodePort, path)
+		hcToCreate, err = gce.ensureHttpHealthCheck(loadBalancerName, path, healthCheckNodePort)
```
if the previous call to GetHttpHealthCheck flaked for some random network connectivity reason throwing an error, and you create it again here, you'll get a StatusConflict and end up retrying (good, because we should've retried the first get anway) but the error message will be wrong. 

Suggest checking the error and returning the right message, and only creating if we need to.

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -729,17 +757,16 @@ func (gce *GCECloud) EnsureLoadBalancer(clusterName string, apiService *api.Serv
-			hc, err = gce.GetHttpHealthCheck(loadBalancerName)
+			hcExisting, err = gce.GetHttpHealthCheck(loadBalancerName)
 			if err != nil && !isHTTPErrorCode(err, http.StatusNotFound) {
 				glog.Infof("Failed to retrieve health check %v:%v", loadBalancerName, err)
```
won't we leak a health check in this case?

bprashanth commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -446,6 +450,9 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
 				}
+			} else {
+				// Delete healthcheck responders, if any, previously listening for this service
+				healthcheck.DeleteServiceListener(serviceName.NamespacedName, 0)
```
are we unconditionally calling this for any/all services? are you sure it doesn't have side-effects for normal services? maybe we should do a periodic background gc instaed?

bprashanth commented on pkg/controller/service/servicecontroller.go:
```diff
@@ -422,10 +422,15 @@ func (s *serviceCache) delete(serviceName string) {
+	if oldServiceNeedsLoadBalancer != newServiceNeedsLoadBalancer {
 		s.eventRecorder.Eventf(newService, api.EventTypeNormal, "Type", "%v -> %v",
 			oldService.Spec.Type, newService.Spec.Type)
 		return true
```
if you're simply returning true when either old or new service needs loadbalancing, does the rest of this function with the events even run?

girishkalele replied:
Yes, the rest of the checks won't be necessary if this change is what we want.

bprashanth replied:
hmm, we still want the events telling us when something changed right?

thockin commented:
@pwittrock In or out for .0 ?

girishkalele commented:
@thockin @bprashanth 

I think I will remove the experimental change that periodically triggers EnsureLoadBalancer till we have enough soak testing. Rest of the commit is essential for 1.4 for not leaking http health checks when services are being created/deleted/modified.

girishkalele commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/cloudprovider/providers/gce/gce.go, line 764 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ67QpxcKfIqkyt-r3-764:-KR-i_cpUo58X-vJ85xm:bk5mnwd) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/cloudprovider/providers/gce/gce.go#L764)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> won't we leak a health check in this case? 
</details>
Yes, if err is not StatusNotFound, then we encountered a GCE API flake and we pass hcExisting=nil to DeleteTargetPool.
We've already deleted the forwarding rule earlier, so I wanted to let it complete the teardown/recreate sequence and not bail out here.
The next time through the sync loop, we will detect the leaked/stale health check and delete it (or attempt to delete it).

---

*[pkg/controller/service/servicecontroller.go, line 436 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ7fjUHK7zAe24o-r3-436:-KR-kOMKdIBAqaGaVkmq:bakpl10) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/controller/service/servicecontroller.go#L436)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> hmm, we still want the events telling us when something changed right? 
</details>
Lets break this change out to another PR - we might impact the other cloudproviders in unexpected ways.

---

*[pkg/proxy/iptables/proxier.go, line 455 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ5p5WtYeiNtsaG-r3-455:-KR-kZHuT-0nNb0qixMe:b-7s29p5) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/proxy/iptables/proxier.go#L455)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> are we unconditionally calling this for any/all services? are you sure it doesn't have side-effects for normal services? maybe we should do a periodic background gc instaed? 
</details>
This is a trivial op - it is discarded very quickly if the map lookup fails in proxy/healthcheck/listener.go

```
        sr, serviceFound := h.serviceResponderMap[req.serviceName]
        if !req.add {
                if !serviceFound {
                        return false
                }
```

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
SG, ping thread when you've made the changes

Commit:
Add NotFound safety check to the HealthCheck reconciler

girishkalele commented:
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/cloudprovider/providers/gce/gce.go, line 720 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ67QpxcKfIqkyt-r3-720:-KR-mKyVerYW3nC1OsaX:b-8b11j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/cloudprovider/providers/gce/gce.go#L720)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> this is dangerous, you could've receive any error and you're concluding that the hc doesn't exist so you need to recreat the targetPool. Suggest checking for explicitly `isHTTPErrorCode(err, http.StatusNotFound)` 
</details>
Added

---

*[pkg/cloudprovider/providers/gce/gce.go, line 730 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ67QpxcKfIqkyt-r3-730:-KR-mMLPZF-SJdIoRYzq:b-8b11j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/cloudprovider/providers/gce/gce.go#L730)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> if the previous call to GetHttpHealthCheck flaked for some random network connectivity reason throwing an error, and you create it again here, you'll get a StatusConflict and end up retrying (good, because we should've retried the first get anway) but the error message will be wrong. 
> 
> 
> 
> Suggest checking the error and returning the right message, and only creating if we need to. 
</details>
Added

---

*[pkg/cloudprovider/providers/gce/gce.go, line 1458 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ67QpxcKfIqkyt-r3-1458:-KR-mQIrEvUnLv68yNo_:b-8b11j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/cloudprovider/providers/gce/gce.go#L1458)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> can you clarify in a comment when/if we come through and cleanup this leak in the case of an error? 
</details>
Added

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Fixed CR comments and removed the experimental commit.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-:-KR-mUAfze6CcJVYzsRb:b-y9vtnh)*
<!-- Sent from Reviewable.io -->

thockin commented:
ping @bprashanth

bprashanth commented:
Oh, sorry, looking now

bprashanth commented:
*[pkg/cloudprovider/providers/gce/gce.go, line 764 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-KQXeBJ67QpxcKfIqkyt-r3-764:-KRFAPlTMRbTkw9E68yU:b-chd7p3) ([raw file](https://github.com/kubernetes/kubernetes/blob/3bd156dee16c8da7745c87e5bf8179da8eb05e16/pkg/cloudprovider/providers/gce/gce.go#L764)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> Yes, if err is not StatusNotFound, then we encountered a GCE API flake and we pass hcExisting=nil to DeleteTargetPool.
> We've already deleted the forwarding rule earlier, so I wanted to let it complete the teardown/recreate sequence and not bail out here.
> The next time through the sync loop, we will detect the leaked/stale health check and delete it (or attempt to delete it).
</details>
please add TODO comment since we're not actually periodically syncing yet (right? )

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828)*
<!-- Sent from Reviewable.io -->

bprashanth commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -719,6 +719,37 @@ func (gce *GCECloud) EnsureLoadBalancer(clusterName string, apiService *api.Serv
+		if err != nil {
+			// This logic exists to detect a transition for a pre-existing service and turn on
+			// the tpNeedsUpdate flag to delete/recreate fwdrule/tpool adding the health check
+			// to the target pool.
```
we should make the implications of toggling the annotation very clear in the alpha limitations section of the official doc (it'll lead to recreating targetpools which causes downtime).

bprashanth commented:
LGTM

bprashanth commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31828#-:-KRFBSiY57R5SpZJUcE3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
The 2 follow up actions: 
1. Make sure we clarify that toggling esipp has delete/create effects on cloudprovider resourcs and can cause downtime
2. add a todo

1. isn't in this pr, 2 can happen in any following pr because the problem is basically: till we periodically resync services with cloudprovider errors at bad times can lead to resource leaks.

pwittrock commented:
1.  Please provide a risk assessment in the description

2. How critical is this?  blocking / important / best effort?  We have dozens of cherrypick requests and each one creates a lot of overhead because large batches introduce a high-risk of destabilizing the tests in a way that is hard to track down.   We have to break them up into smaller batches and monitor the tests to catch issues early on and track them back to their root cause.

bprashanth commented:
>Please provide a risk assessment in the description

medium, scoped to services with an alpha feature expressed via annoation. Risk is that we affect normal service behavior somewhere.

> How critical is this? blocking / important / best effort? 

important. The feature in question is p0 for 1.4, and this pr plugs a resource leak.  

@girishkalele might add clarifications as the author.

thockin commented:
To be devil's advocate - a resource leak on an alpha feature is NOT a p0.
This could push to 1.4.1.  I will leave this up to @pwittrock.

On Sat, Sep 10, 2016 at 2:51 PM, Prashanth B <notifications@github.com>
wrote:

> Please provide a risk assessment in the description
>
> medium, scoped to services with an alpha feature expressed via annoation.
> Risk is that we affect normal service behavior somewhere.
>
> How critical is this? blocking / important / best effort?
>
> important. The feature in question is p0 for 1.4, and this pr plugs a
> resource leak.
>
> @girishkalele <https://github.com/girishkalele> might add clarifications
> as the author.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31828#issuecomment-246143492>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVLEnU-muJukeHVS8NRuek3--nRmwks5qoyZMgaJpZM4JyHrV>
> .
>

pwittrock commented:
Its already in :)

bprashanth commented:
I mean, it's going into a beta.1 not the final release. We still have room to fix a bug in the bug fix :)
